[Frontend] Allow -emit-interface with -typecheck

jrose-apple opened the pull request:
Textual module interfaces don't actually depend on SILGen, so we shouldn't need to run SILGen (or serialize an entire binary module) if we're just trying to emit a textual interface. On the other hand, if we *are* going to run SILGen and then SIL diagnostics, we shouldn't delay those diagnostics by spending time emitting a textual interface, or for that matter a TBD file.

Using this, update all the ModuleInterface tests that use `-emit-module -o /dev/null` to use `-typecheck` instead, except for those using `-merge-modules`.

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
@swift-ci Please test source compatibility

harlanhaskins commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -867,6 +867,40 @@ emitIndexData(CompilerInvocation &Invocation, CompilerInstance &Instance) {
+/// These are extracted out so that they can be invoked early when using
+/// `-typecheck`, but skipped for any mode that runs SIL diagnostics if there's
+/// an error found there (to get those diagnostics back to the user faster).
+static bool emitAnyWholeModulePostTypeCheckSupplementaryOutputs(
```
This is much nicer, thank you!

harlanhaskins commented on test/ModuleInterface/stdlib.swift:
```diff
@@ -1,3 +1,4 @@
+// RUN: %target-swift-frontend -typecheck -emit-interface-path - -parse-stdlib %s | %FileCheck %s
 // RUN: %target-swift-frontend -emit-interface-path - -emit-module -o /dev/null -parse-stdlib %s | %FileCheck %s
```
I think this line can be removed

jrose-apple replied:
Ah, I wanted to have at least one test that did things both ways; a -parse-stdlib test seemed like a good example.

harlanhaskins replied:
Sounds good, then!

jrose-apple commented:
Aha, yes, the very thing I changed was, in fact, being tested.

jrose-apple committed:
[Frontend] Allow -emit-interface with -typecheck

Textual module interfaces don't actually depend on SILGen, so we
shouldn't need to run SILGen (or serialize an entire binary module) if
we're just trying to emit a textual interface. On the other hand, if
we /are/ going to run SILGen and then SIL diagnostics, we shouldn't
delay those diagnostics by spending time emitting a textual interface,
or for that matter a TBD file.

Using this, update all the ModuleInterface tests that use
`-emit-module -o /dev/null` to use `-typecheck` instead, except for
those using `-merge-modules`.

jrose-apple commented:
[Debug source compat](https://ci.swift.org/job/swift-PR-source-compat-suite-debug/200/), [Release source compat](https://ci.swift.org/job/swift-PR-source-compat-suite/1972/).

@swift-ci Please smoke test
